app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,145 @@
+    public HereFunctionHandler() {}
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    // Setting up and subscribing to the location manager is delayed until here() is actually
+    // evaluated.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if(!locationUpdatesRequested) requestLocationUpdates();
+        if(location == null) return "";
+        return location.getDisplayText();
+    }
+
+    private void requestLocationUpdates() {
+        mLocationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
+        Set<String> mProviders = GeoUtils.evaluateProviders(mLocationManager);
+
+        for (String provider : mProviders) {
+            if ((provider.equals(LocationManager.GPS_PROVIDER) && ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) ||
+                    (provider.equals(LocationManager.NETWORK_PROVIDER) && ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED)) {
+                // This is non-null if the calling activity pauses.
+                if (location == null) {
+                    Location lastKnownLocation = mLocationManager.getLastKnownLocation(provider);
+                    this.location = toGeoPointData(lastKnownLocation);
+                    Log.i("HereFunctionHandler", "last known location: " + this.location.getDisplayText());
+                }
+
+                // Looper is necessary because requestLocationUpdates is called inside an AsyncTask (EntityLoaderTask).
+                // What values for minTime and minDistance?
+                mLocationManager.requestLocationUpdates(provider, 0, 0, this, Looper.getMainLooper());
+            }
+        }
+
+        locationUpdatesRequested = true;
+    }
+
+    // Clients must call this when done using handler.
+    public void stopLocationUpdates() {
+        // stops the GPS. Note that this will turn off the GPS if the screen goes to sleep.
+        if (ContextCompat.checkSelfPermission(this.context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
+                ContextCompat.checkSelfPermission(this.context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
+            mLocationManager.removeUpdates(this);
+        }
+
+        locationUpdatesRequested = false;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        // Do we need to check the accuracy of the location?
+        this.location = toGeoPointData(location);
+        Log.i("HereFunctionHandler", "location has been set to " + this.location.getDisplayText());
+        // Trigger refreshing of entity list.
+        if (entitySelectActivity != null) {
+            entitySelectActivity.loadEntities();
```

Comment:
Might eventually want to only do this reloading when the final location is acquired. That is, refresh only when we have acquired a satisfactorily accurate location.